File: LeetCodePractice/src/DFS/NumbersWithSameConsecutiveDifferences.java
package DFS;

import java.util.ArrayList;
import java.util.List;

/*
Return all non-negative integers of length N such that the absolute difference between every two consecutive digits is K.

Note that every number in the answer must not have leading zeros except for the number 0 itself. For example, 01 has one leading zero and is invalid, but 0 is valid.

You may return the answer in any order.



Example 1:

Input: N = 3, K = 7
Output: [181,292,707,818,929]
Explanation: Note that 070 is not a valid number, because it has leading zeroes.
Example 2:

Input: N = 2, K = 1
Output: [10,12,21,23,32,34,43,45,54,56,65,67,76,78,87,89,98]


Note:

1 <= N <= 9
0 <= K <= 9
 */
public class NumbersWithSameConsecutiveDifferences {
    public int[] numsSameConsecDiff(int N, int K) {
        List<Integer> res = new ArrayList<>();
        if(N == 1){
            res.add(0);
        }
        dfs(res, 0, N, K);
        return res.stream().mapToInt(i -> i).toArray();
    }

    private void dfs(List<Integer> res, int cur, int n, int k) {
        if(n == 0){
            res.add(cur);
            return;
        }
        for(int d = 0; d <= 9; d++){
            if(cur == 0){
                if(d != 0){
                    dfs(res, d, n - 1, k);
                }
            } else {
                int lastDigit = cur % 10;
                if(Math.abs(lastDigit - d) == k){
                    dfs(res, cur * 10 + d, n - 1, k);
                }
            }
        }
    }
}

File: LeetCodePython/NumbersWithSameConsecutiveDifferences.py
"""
Return all non-negative integers of length n such that the absolute difference between every two consecutive digits is k.

Note that every number in the answer must not have leading zeros. For example, 01 has one leading zero and is invalid.

You may return the answer in any order.

Example 1:
Input: n = 3, k = 7
Output: [181,292,707,818,929]
Explanation: Note that 070 is not a valid number, because it has leading zeroes.

Example 2:
Input: n = 2, k = 1
Output: [10,12,21,23,32,34,43,45,54,56,65,67,76,78,87,89,98]

Constraints:
2 <= n <= 9
0 <= k <= 9
"""
from typing import List


class NumbersWithSameConsecutiveDifferences:
    def numsSameConsecDiff(self, n: int, k: int) -> List[int]:
        res = []
        if n == 1:
            res.append(1)

        def dfs(num, i):
            if i == 0:
                res.append(num)
                return
            for d in range(0, 10):
                if num == 0:
                    if d != 0:
                        dfs(d, i - 1)
                else:
                    last_digit = num % 10
                    if abs(last_digit - d) == k:
                        dfs(num * 10 + d, i - 1)

        dfs(0, n)
        return res
